crates/provider/src/layers/nonce.rs

```diff
@@ -0,0 +1,207 @@
+use async_trait::async_trait;
+use dashmap::{mapref::entry::Entry, DashMap};
+use std::marker::PhantomData;
+
+/// A layer that fills nonces on transactions.
+///
+/// The layer will fetch the transaction count for any new account it sees, store it locally and
+/// increment the locally stored nonce as transactions are sent via [`Provider::send_transaction`].
+///
+/// If you use layers that redirect the behavior of [`Provider::send_transaction`] (e.g.
+/// [`SignerLayer`]), you should add this layer before those.
+///
+/// # Note
+///
+/// - If the transaction request does not have a sender set, this layer will not fill nonces.
+/// - Using two providers with their own nonce layer can potentially fill invalid nonces if
+///   transactions are sent from the same address, as the next nonce to be used is cached internally
+///   in the layer.
+///
+/// # Example
+///
+/// ```rs
+/// # async fn test<T: Transport + Clone, S: NetworkSigner<Ethereum>>(transport: T, signer: S) {
+/// let provider = ProviderBuilder::<_, Ethereum>::new()
+///     .layer(ManagedNonceLayer)
+///     .signer(EthereumSigner::from(signer)) // note the order!
+///     .network::<Ethereum>()
+///     .provider(RootProvider::new(transport));
+///
+/// provider.send_transaction(TransactionRequest::default()).await;
+/// # }
+/// ```
+///
+/// [`SignerLayer`]: crate::layers::SignerLayer
+pub struct ManagedNonceLayer;
+
+impl<P, N, T> ProviderLayer<P, N, T> for ManagedNonceLayer
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    type Provider = ManagedNonceProvider<N, T, P>;
+
+    fn layer(&self, inner: P) -> Self::Provider {
+        ManagedNonceProvider { inner, nonces: DashMap::default(), _phantom: PhantomData }
+    }
+}
+
+/// A provider that manages account nonces.
+///
+/// Fills nonces for transaction requests if unset.
+///
+/// # Note
+///
+/// If the transaction requests do not have a sender set, this provider will not set nonces.
+///
+/// You cannot construct this provider directly. Use [`ProviderBuilder`] with a
+/// [`ManagedNonceLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+pub struct ManagedNonceProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    nonces: DashMap<Address, u64>,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> ManagedNonceProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    async fn get_nonce(&self, from: Address) -> TransportResult<U64> {
+        let nonce = match self.nonces.entry(from) {
+            Entry::Occupied(entry) => {
+                let next_nonce = entry.get() + 1;
+                entry.replace_entry(next_nonce);
+                next_nonce
+            }
+            Entry::Vacant(entry) => {
+                let nonce: u64 = self.inner.get_transaction_count(from, None).await?.to();
+                entry.insert(nonce);
```

Comment:
this uses a sync lock across an await point, which can cause issues,

imo a better approach here would be a sync lock over an async lock that guards the nonce